File: Algorithms/1423_Maximum_Points_You_Can_Obtain_from_Cards/Java/MaximumPointsYouCanObtainFromCardsSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public int maxScore(int[] cardPoints, int k) {
        int length = cardPoints.length;
        if (k == length) return Arrays.stream(cardPoints).sum();

        int left = 0, right = length - k;
        int minSum = Arrays.stream(Arrays.copyOfRange(cardPoints, left, right)).sum();
        int curSum = Arrays.stream(Arrays.copyOfRange(cardPoints, left, right)).sum();

        while (right < length) {
            curSum = curSum - cardPoints[left] + cardPoints[right];
            minSum = Math.min(curSum, minSum);
            left++;
            right++;
        }
        return Arrays.stream(cardPoints).sum() - minSum;
    }
}

File: Algorithms/1423_Maximum_Points_You_Can_Obtain_from_Cards/Python/Maximum_Points_You_Can_Obtain_from_Cards_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def maxScore(self, cardPoints, k):

        length = len(cardPoints)
        if length == k: return sum(cardPoints)

        left, right = 0, length - k
        min_sum = cur_sum = sum(cardPoints[left:right])

        while right < length:
            cur_sum = cur_sum - cardPoints[left] + cardPoints[right]
            min_sum = min(min_sum, cur_sum)
            left += 1
            right += 1

        return sum(cardPoints) - min_sum





